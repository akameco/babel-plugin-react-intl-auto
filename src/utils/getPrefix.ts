import { relative, dirname, sep } from 'path'
import { State } from '../types'
import { dotPath } from '.'

const escapeRegExp = (text: string) => {
  return text.replace(/[-[\]{}()*+?.,\\^$|#\s]/gu, '\\$&')
}

const dotPathReplace = (
  fomatted: string,
  removePrefix: string,
  separator?: string
) => {
  const exp = `^${removePrefix.replace(/\//gu, '')}\\${dotPath(
    sep,
    separator
  )}?`
  let reg: RegExp

  // certain separators can throw an error and need to be escaped
  // e.g. "_"
  try {
    reg = new RegExp(exp, 'u')
  } catch (error) {
    reg = new RegExp(escapeRegExp(exp), 'u')
  }

  return dotPath(fomatted, separator).replace(reg, '')
}

export const getPrefix = (
  {
    file: {
      opts: { filename },
    },
    opts: { removePrefix, filebase = false, separator },
  }: State,
  exportName: string | null
) => {
  if (removePrefix === true) {
    return exportName === null ? '' : exportName
  }
  const file = relative(process.cwd(), filename)
  const fomatted = filebase ? file.replace(/\..+$/u, '') : dirname(file)
  removePrefix =
    removePrefix === undefined || removePrefix === false ? '' : removePrefix

  const fixed =
    removePrefix instanceof RegExp
      ? dotPath(fomatted.replace(removePrefix, ''), separator)
      : dotPathReplace(fomatted, removePrefix, separator)

  if (exportName === null) {
    return fixed
  }

  if (fixed === '') {
    return exportName
  }

  return `${fixed}.${exportName}`
}
