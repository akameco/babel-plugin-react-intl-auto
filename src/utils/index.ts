import { sep } from 'path'
import * as t from '@babel/types'
import murmur from 'murmurhash3js'
import { NodePath } from '@babel/core'

const REG = new RegExp(`\\${sep}`, 'gu')

const isObjectProperties = (
  properties: NodePath[]
): properties is NodePath<t.ObjectProperty>[] =>
  properties.every(p => p.isObjectProperty())

export const createHash = (message: string) => `${murmur.x86.hash32(message)}`

export const dotPath = (str: string, separator: string | undefined = '.') =>
  str.replace(REG, separator)

export const objectProperty = (
  key: string,
  value: string | t.StringLiteral | t.TemplateLiteral
) => {
  const valueNode = typeof value === 'string' ? t.stringLiteral(value) : value
  return t.objectProperty(t.stringLiteral(key), valueNode)
}

export function getObjectProperties(path: NodePath) {
  if (path.isObjectExpression()) {
    const properties = path.get('properties')
    if (isObjectProperties(properties)) {
      return properties
    }
  } else if (path.isIdentifier()) {
    const binding = path.scope.getBinding(path.node.name)
    if (!binding) {
      return null
    }
    const properties = binding.path.get('init.properties')
    if (Array.isArray(properties) && isObjectProperties(properties)) {
      return properties
    }
  }
  return null
}
