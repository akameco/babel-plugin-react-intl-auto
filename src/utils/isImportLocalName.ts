import { NodePath } from '@babel/core'
import * as t from '@babel/types'
import { State } from '../types'

export const isImportLocalName = (
  name: string | null | undefined,
  allowedNames: string[],
  { file, opts: { moduleSourceName = 'react-intl' } }: State
) => {
  const isSearchedImportSpecifier = (
    specifier: NodePath<
      t.ImportDefaultSpecifier | t.ImportNamespaceSpecifier | t.ImportSpecifier
    >
  ) =>
    specifier.isImportSpecifier() &&
    allowedNames.includes(specifier.node.imported.name) &&
    (!name || specifier.node.local.name === name)

  let isImported = false

  if (file && file.path) {
    file.path.traverse({
      ImportDeclaration: {
        exit(path) {
          const specifiers = path.get('specifiers') as NodePath<
            | t.ImportDefaultSpecifier
            | t.ImportNamespaceSpecifier
            | t.ImportSpecifier
          >[]
          isImported =
            path.node.source.value.includes(moduleSourceName) &&
            specifiers.some(specifier => isSearchedImportSpecifier(specifier))

          if (isImported) {
            path.stop()
          }
        },
      },
    })
  }

  return isImported
}
