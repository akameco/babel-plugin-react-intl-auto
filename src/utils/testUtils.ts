// eslint-disable-next-line import/no-extraneous-dependencies
import pluginTester from 'babel-plugin-tester'
import { Opts } from '../types'
import plugin from '..'

export function cases(
  filename: string,
  testCases: {
    title: string
    tests: { title: string; code: string }[]
    pluginOptions?: Opts
  }[]
) {
  const defaultOpts = {
    title: '',
    plugin,
    snapshot: true,
    babelOptions: { filename, parserOpts: { plugins: ['jsx'] } },
    formatResult: (r: any) => r,
    tests: [],
  }

  for (const testCase of testCases) {
    testCase.tests = testCase.tests.map(t => ({ ...t, title: t.title }))
    pluginTester({ ...defaultOpts, ...testCase })
  }
}
