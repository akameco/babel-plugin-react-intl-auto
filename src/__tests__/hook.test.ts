import path from 'path'
import { cases } from '../utils/testUtils'

const filename = path.resolve(__dirname, '..', '__fixtures__', 'messages.js')

const defaultTest = {
  title: 'with useIntl hook imported',
  code: `
import { useIntl } from 'react-intl';

intl.formatMessage({ defaultMessage: "hello" });
  `,
}

const multiUseTest = {
  title: 'multiple uses',
  code: `
import { useIntl } from 'react-intl';

intl.formatMessage({ defaultMessage: "hello" });
intl.formatMessage({ defaultMessage: "hello" });
intl.formatMessage({ defaultMessage: "some other message" });
`,
}

const withValueInMessageTest = {
  title: 'with a value interpolated in the message',
  code: `
import { useIntl } from 'react-intl';

intl.formatMessage({ defaultMessage: \`template string ${1 + 1}\` });
`,
}

const withVariableMessageTest = {
  title: 'with a variable as the defaultMessage',
  code: `
import { useIntl } from 'react-intl';

const message = "variable message";

intl.formatMessage({ defaultMessage: message });
`,
}

const withVariableMessageDescriptor = {
  title: 'with a variable as the defaultMessage',
  code: `
import { useIntl } from 'react-intl';
import { message } from './messages'

intl.formatMessage(messages);
`,
}

const withCustomProperties = {
  title: 'with custom properties in formatMessage call',
  code: `
import { useIntl } from 'react-intl';

intl.formatMessage({ defaultMessage: "custom prop", other: 123 });
`,
}

const someSupportedUseCases = {
  title: 'some supported use cases',
  code: `
import { useIntl } from 'react-intl';

const Component2 = () => {
  const intl = useIntl();
  const label = intl.formatMessage({ defaultMessage: "hello" });
  return (
    <button aria-label={intl.formatMessage({ defaultMessage: "hello" })}>
      {intl.formatMessage({ defaultMessage: "hello" })}
    </button>
  );
};
  `,
}

const importAsTest = {
  title: 'with FormattedMessage imported as something else',
  code: `
import { useIntl as i18n } from 'react-intl';

intl.formatMessage({ defaultMessage: "i18n" });
`,
}

const throwWhenNotAnalyzableTest = {
  title: 'throws if defaultMessage isnâ€™t analyzable',
  code: `
import { useIntl } from 'react-intl';
const getMsg = () => 'hello';
intl.formatMessage({
  defaultMessage: getMsg(),
});
`,
  error: /\[React Intl Auto\] defaultMessage must be statically evaluate-able for extraction/u,
  snapshot: false,
}

const notTransformIfNotImportedTest = {
  title: 'does nothing if react-intl is not imported',
  snapshot: false,
  code: `
import any from 'any-module';
intl.formatMessage({
  defaultMessage: "hello"
});
`,
}

const notTransformIfIdIsProvided = {
  title: 'does nothing if id is already provided',
  snapshot: false,
  code: `
import { useIntl } from 'react-intl';
intl.formatMessage({
  id: "my.custom.id",
  defaultMessage: "hello"
});
`,
}

const injectIntlWithProps = {
  title: 'with injectIntl',
  code: `
import { injectIntl } from 'react-intl';
function App({ intl }) {
  return <div>{intl.formatMessage({ defaultMessage: 'hello' })}</div>
}

export default injectIntl(App)
  `,
}

const withKeyFlag = {
  title: 'withKeyFlag',
  code: `
import { useIntl } from 'react-intl';
intl.formatMessage({
  key: 'foobar',
  defaultMessage: 'hello'
});
  `,
}

const tests = [
  defaultTest,
  multiUseTest,
  withValueInMessageTest,
  withVariableMessageTest,
  withVariableMessageDescriptor,
  withCustomProperties,
  someSupportedUseCases,
  importAsTest,
  throwWhenNotAnalyzableTest,
  notTransformIfNotImportedTest,
  notTransformIfIdIsProvided,
  injectIntlWithProps,
  withKeyFlag,
]

cases(filename, [
  { title: 'default', tests },
  {
    title: 'removePrefix = "src"',
    tests: [defaultTest],
    pluginOptions: { removePrefix: 'src' },
  },

  {
    title: 'removePrefix = "src/" -- with slash',
    tests: [defaultTest],
    pluginOptions: { removePrefix: 'src/' },
  },

  {
    title: 'filebase = true',
    tests: [defaultTest],
    pluginOptions: { filebase: true },
  },

  // {
  //   title: 'includeExportName = true',
  //   tests: [defaultTest],
  //   pluginOptions: { includeExportName: true },
  // },
  // {
  //   title: 'includeExportName = all',
  //   tests: [defaultTest],
  //   pluginOptions: { includeExportName: 'all' },
  // },
  // {
  //   title: 'removePrefix = true, includeExportName = true',
  //   tests: [defaultTest],
  //   pluginOptions: { removePrefix: true, includeExportName: true },
  // },
  {
    title: 'removePrefix = false',
    tests: [defaultTest],
    pluginOptions: { removePrefix: false },
  },

  // {
  //   title: 'removePrefix = true, includeExportName = all',
  //   tests: [defaultTest],
  //   pluginOptions: { removePrefix: true, includeExportName: 'all' },
  // },
  // {
  //   title: 'extractComments = false',
  //   tests: [defaultTest],
  //   pluginOptions: { extractComments: false },
  // },
  {
    title: 'removePrefix = /__fixtures__/',
    tests: [defaultTest],
    pluginOptions: { removePrefix: /[\\/]__fixtures__/u },
  },

  {
    title: 'removePrefix = "src.__fixtures__"',
    tests: [defaultTest],
    pluginOptions: { removePrefix: 'src.__fixtures__' },
  },
  {
    title: 'useKey = true',
    tests: [defaultTest, withKeyFlag],
    pluginOptions: { useKey: true },
  },
  {
    title: 'removeDefaultMessage = true',
    tests: [defaultTest],
    pluginOptions: { removeDefaultMessage: true },
  },
])
