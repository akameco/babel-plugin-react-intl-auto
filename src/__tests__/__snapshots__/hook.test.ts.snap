// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default multiple uses: multiple uses 1`] = `

import { useIntl } from 'react-intl';

intl.formatMessage({ defaultMessage: "hello" });
intl.formatMessage({ defaultMessage: "hello" });
intl.formatMessage({ defaultMessage: "some other message" });

      ↓ ↓ ↓ ↓ ↓ ↓

import { useIntl } from 'react-intl';
intl.formatMessage({
  defaultMessage: "hello",
  "id": "src.__fixtures__.613153351"
});
intl.formatMessage({
  defaultMessage: "hello",
  "id": "src.__fixtures__.613153351"
});
intl.formatMessage({
  defaultMessage: "some other message",
  "id": "src.__fixtures__.831373691"
});

`;

exports[`default some supported use cases: some supported use cases 1`] = `

import { useIntl } from 'react-intl';

const Component2 = () => {
  const intl = useIntl();
  const label = intl.formatMessage({ defaultMessage: "hello" });
  return (
    <button aria-label={intl.formatMessage({ defaultMessage: "hello" })}>
      {intl.formatMessage({ defaultMessage: "hello" })}
    </button>
  );
};

      ↓ ↓ ↓ ↓ ↓ ↓

import { useIntl } from 'react-intl';

const Component2 = () => {
  const intl = useIntl();
  const label = intl.formatMessage({
    defaultMessage: "hello",
    "id": "src.__fixtures__.613153351"
  });
  return <button aria-label={intl.formatMessage({
    defaultMessage: "hello",
    "id": "src.__fixtures__.613153351"
  })}>
      {intl.formatMessage({
      defaultMessage: "hello",
      "id": "src.__fixtures__.613153351"
    })}
    </button>;
};

`;

exports[`default with FormattedMessage imported as something else: with FormattedMessage imported as something else 1`] = `

import { useIntl as i18n } from 'react-intl';

intl.formatMessage({ defaultMessage: "i18n" });

      ↓ ↓ ↓ ↓ ↓ ↓

import { useIntl as i18n } from 'react-intl';
intl.formatMessage({
  defaultMessage: "i18n",
  "id": "src.__fixtures__.94348014"
});

`;

exports[`default with a value interpolated in the message: with a value interpolated in the message 1`] = `

import { useIntl } from 'react-intl';

intl.formatMessage({ defaultMessage: \`template string 2\` });

      ↓ ↓ ↓ ↓ ↓ ↓

import { useIntl } from 'react-intl';
intl.formatMessage({
  defaultMessage: \`template string 2\`,
  "id": "src.__fixtures__.1045198380"
});

`;

exports[`default with a variable as the defaultMessage: with a variable as the defaultMessage 1`] = `

import { useIntl } from 'react-intl';

const message = "variable message";

intl.formatMessage({ defaultMessage: message });

      ↓ ↓ ↓ ↓ ↓ ↓

import { useIntl } from 'react-intl';
const message = "variable message";
intl.formatMessage({
  defaultMessage: message,
  "id": "src.__fixtures__.3082794952"
});

`;

exports[`default with a variable as the defaultMessage: with a variable as the defaultMessage 2`] = `

import { useIntl } from 'react-intl';
import { message } from './messages'

intl.formatMessage(messages);

      ↓ ↓ ↓ ↓ ↓ ↓

import { useIntl } from 'react-intl';
import { message } from './messages';
intl.formatMessage(messages);

`;

exports[`default with custom properties in formatMessage call: with custom properties in formatMessage call 1`] = `

import { useIntl } from 'react-intl';

intl.formatMessage({ defaultMessage: "custom prop", other: 123 });

      ↓ ↓ ↓ ↓ ↓ ↓

import { useIntl } from 'react-intl';
intl.formatMessage({
  defaultMessage: "custom prop",
  "id": "src.__fixtures__.2983810267",
  other: 123
});

`;

exports[`default with injectIntl: with injectIntl 1`] = `

import { injectIntl } from 'react-intl';
function App({ intl }) {
  return <div>{intl.formatMessage({ defaultMessage: 'hello' })}</div>
}

export default injectIntl(App)

      ↓ ↓ ↓ ↓ ↓ ↓

import { injectIntl } from 'react-intl';

function App({
  intl
}) {
  return <div>{intl.formatMessage({
      defaultMessage: 'hello',
      "id": "src.__fixtures__.613153351"
    })}</div>;
}

export default injectIntl(App);

`;

exports[`default with useIntl hook imported: with useIntl hook imported 1`] = `

import { useIntl } from 'react-intl';

intl.formatMessage({ defaultMessage: "hello" });

      ↓ ↓ ↓ ↓ ↓ ↓

import { useIntl } from 'react-intl';
intl.formatMessage({
  defaultMessage: "hello",
  "id": "src.__fixtures__.613153351"
});

`;

exports[`default withKeyFlag: withKeyFlag 1`] = `

import { useIntl } from 'react-intl';
intl.formatMessage({
  key: 'foobar',
  defaultMessage: 'hello'
});

      ↓ ↓ ↓ ↓ ↓ ↓

import { useIntl } from 'react-intl';
intl.formatMessage({
  key: 'foobar',
  defaultMessage: 'hello',
  "id": "src.__fixtures__.613153351"
});

`;

exports[`filebase = true with useIntl hook imported: with useIntl hook imported 1`] = `

import { useIntl } from 'react-intl';

intl.formatMessage({ defaultMessage: "hello" });

      ↓ ↓ ↓ ↓ ↓ ↓

import { useIntl } from 'react-intl';
intl.formatMessage({
  defaultMessage: "hello",
  "id": "src.__fixtures__.messages.613153351"
});

`;

exports[`removeDefaultMessage = true with useIntl hook imported: with useIntl hook imported 1`] = `

import { useIntl } from 'react-intl';

intl.formatMessage({ defaultMessage: "hello" });

      ↓ ↓ ↓ ↓ ↓ ↓

import { useIntl } from 'react-intl';
intl.formatMessage({
  "id": "src.__fixtures__.613153351"
});

`;

exports[`removePrefix = "src" with useIntl hook imported: with useIntl hook imported 1`] = `

import { useIntl } from 'react-intl';

intl.formatMessage({ defaultMessage: "hello" });

      ↓ ↓ ↓ ↓ ↓ ↓

import { useIntl } from 'react-intl';
intl.formatMessage({
  defaultMessage: "hello",
  "id": "__fixtures__.613153351"
});

`;

exports[`removePrefix = "src.__fixtures__" with useIntl hook imported: with useIntl hook imported 1`] = `

import { useIntl } from 'react-intl';

intl.formatMessage({ defaultMessage: "hello" });

      ↓ ↓ ↓ ↓ ↓ ↓

import { useIntl } from 'react-intl';
intl.formatMessage({
  defaultMessage: "hello",
  "id": "613153351"
});

`;

exports[`removePrefix = "src/" -- with slash with useIntl hook imported: with useIntl hook imported 1`] = `

import { useIntl } from 'react-intl';

intl.formatMessage({ defaultMessage: "hello" });

      ↓ ↓ ↓ ↓ ↓ ↓

import { useIntl } from 'react-intl';
intl.formatMessage({
  defaultMessage: "hello",
  "id": "__fixtures__.613153351"
});

`;

exports[`removePrefix = /__fixtures__/ with useIntl hook imported: with useIntl hook imported 1`] = `

import { useIntl } from 'react-intl';

intl.formatMessage({ defaultMessage: "hello" });

      ↓ ↓ ↓ ↓ ↓ ↓

import { useIntl } from 'react-intl';
intl.formatMessage({
  defaultMessage: "hello",
  "id": "src.613153351"
});

`;

exports[`removePrefix = false with useIntl hook imported: with useIntl hook imported 1`] = `

import { useIntl } from 'react-intl';

intl.formatMessage({ defaultMessage: "hello" });

      ↓ ↓ ↓ ↓ ↓ ↓

import { useIntl } from 'react-intl';
intl.formatMessage({
  defaultMessage: "hello",
  "id": "src.__fixtures__.613153351"
});

`;

exports[`useKey = true with useIntl hook imported: with useIntl hook imported 1`] = `

import { useIntl } from 'react-intl';

intl.formatMessage({ defaultMessage: "hello" });

      ↓ ↓ ↓ ↓ ↓ ↓

import { useIntl } from 'react-intl';
intl.formatMessage({
  defaultMessage: "hello",
  "id": "src.__fixtures__.613153351"
});

`;

exports[`useKey = true withKeyFlag: withKeyFlag 1`] = `

import { useIntl } from 'react-intl';
intl.formatMessage({
  key: 'foobar',
  defaultMessage: 'hello'
});

      ↓ ↓ ↓ ↓ ↓ ↓

import { useIntl } from 'react-intl';
intl.formatMessage({
  key: 'foobar',
  defaultMessage: 'hello',
  "id": "src.__fixtures__.foobar"
});

`;
